### ISSUE TITLE
In addition to CARGO_TARGET_DIR env var, support --target-dir command line option

### ISSUE THREAD

**Issue by matklad** [rust-lang/cargo#5308]
@nirbheek you wanted this for messon integration. Could you elaborate on why do you think it is better than an environmet variable?

@rust-lang/cargo it seems that such addition would be E-easy and straightforward? The only complication I can think of are build scripts which invoke cargo recursively for whatever reason, but we can just set the env car for them manually. Should we implement this? If yes, I’ll write some mentoring instructions :)

**Comment by alexcrichton** [comment]
Seems reasonable to me to implement!

**Comment by nirbheek** [comment]
> Could you elaborate on why do you think it is better than an environment variable?

Speaking generally, in my opinion command-line arguments are more explicit and user-friendly than environment variables. They are easier to change or discover, and are less likely to have unintended side-effects.

**Comment by matklad** [comment]
So, if anyone wants to implement this, here are some pointers to code:

1) To specify command line argment itself, you should define a helper function [over here](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/bin/command_prelude.rs#L108) and than call it for all relevant commands. I think that all commands that call `arg_target_triple` or `arg_jobs` would probably want to call `arg_target_dir` as well

2) The code to calculate `target_dir` is [here](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/cargo/util/config.rs#L240)

3) I think that probably we can thread value of command line argument to config via [this call](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/bin/cli.rs#L75)?

4) The test for target dir is [here](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/tests/testsuite/build.rs#L3619)

**Comment by smithsps** [comment]
@matklad I would be interested in working on this. What would be the best way to direct questions?

**Comment by matklad** [comment]
@smithsps awesome! 

You can ask the questions right here, and we have [an IRC channel](https://kiwiirc.com/client/irc.mozilla.org/cargo) as well!

**Comment by smithsps** [comment]
The [configure](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/cargo/util/config.rs#L455) function can be used to pass our target_dir String, and work with it in config. My current approach in config was:

Adding a target directory to the [config struct](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/cargo/util/config.rs#L39).
```rust
target_dir: Option<Filesystem>,
```
Changing the [target_dir()](https://github.com/rust-lang/cargo/blob/e837ce064fcf28d5f739f87376384c18a33c6c69/src/cargo/util/config.rs#L240) to:

```rust
    pub fn target_dir(&self) -> &Option<Filesystem> {
        &self.target_dir
    }

    fn get_target_dir(&self, target_dir_param: Option<String>) -> CargoResult<Option<Filesystem>> {
        if let Some(dir) = target_dir_param {
            Ok(Some(Filesystem::new(self.cwd.join(dir))))
        } else if let Some(dir) = env::var_os("CARGO_TARGET_DIR") {
            Ok(Some(Filesystem::new(self.cwd.join(dir))))
        } else if let Some(val) = self.get_path("build.target-dir")? {
            let val = self.cwd.join(val.val);
            Ok(Some(Filesystem::new(val)))
        } else {
            Ok(None)
        }
    }
```
And adding this to configure():
```rust
let target_dir = self.get_target_dir(target_dir)?;
```
But, I had a couple of questions with this approach:
1. It doesn't seem 'correct' to be adding a field to struct, when normally `target_dir` generates a new Filesystem value every time? However, `frozen` and `locked` are passed this way. Would this be OK to also store our `target_dir`, in this `config` struct? 
2. Is configure() called in every case? This solution wouldn't cover cases that don't. Would the various unit tests be the best way to figure that out?
3. Is `&Option<Filesystem>` proper rusty? It seems wrong. Maybe something like `self.target_dir.as_ref()` and `Option<&Filesystem>`.

**Comment by smithsps** [comment]
Do we only need to create a single test in build that accompanies the existing one? Or do we need one with every command we include arg_target_dir?

**Comment by matklad** [comment]
A single dest would be enogh!

**Comment by smithsps** [comment]
To check in, I've got a working solution.
But, I'm not sure its a complete architectural fit. Could you check if [my approach](https://github.com/smithsps/cargo/commit/c36b0d966f744e6c3c6d8796b0b3ed6b05906f46) is correct?

I'll probably pull request shortly, but I still need to double check the correct subcommands and their various description.

**Comment by dwijnand** [comment]
Fixed by #5393

**Comment by alexcrichton** [comment]
Indeed, thanks!
